rfc/cart-api-v1.md

```diff
@@ -0,0 +1,251 @@
+    if id == 'default':
+        cart = cart_repository.get_default(user_id = bearer_token)
+
+        if not cart:
+            # Either this or: cart = cart_repository.create(user_id = bearer_token, cart_name = '')
+            cart = cart_repository.create(user_id = bearer_token, cart_name = '', default = True)
+    else:
+        cart = cart_repository.get(id)
+
+    return dict(id=cart.id,
+                cart_name=cart.cart_name,
+                entity_type=cart.entity_type,
+                default=cart.default,
+                items=cart_item_repository.find_many_by_cart_id(cart.id))
+```
+
+Here is the sample request.
+
+```
+GET /resources/carts/default
+Authorization: Bearer 261e1aea-82c9-4e38-a879-e2e02ef236a5
+```
+
+Here is the sample response.
+
+```javascript
+{
+    id: "6b801de6-11d4-46c0-aed1-ca2110aa7d3c"
+    cart_name: "",
+    default: true,
+    items: [...]
+}
+```
+
+#### `PUT /resources/carts/{id}`
+
+Here is the request body.
+
+```javascript
+{
+    cart_name: str // Required
+}
+```
+
+and the response is basically the JSON representation of `Cart`.
+
+#### `DELETE /resources/carts/{id}`
+
+* The endpoint will delete the target cart and respond with **HTTP 200** only if the cart belongs to the user making the request.
+  * Otherwise, the endpoint will respond with **HTTP 403**.
+* The endpoint will respond **HTTP 404** if the cart is not found.
+
+### CRUD APIs for `CartItem`
+
+| Method | Path | Description |
+| --- | --- | --- |
+| `POST` | `/resources/carts/{cart_id}/items/` | Create a new cart item |
+| `DELETE` | `/resources/carts/{cart_id}/items/{id}` | Delete a single cart item |
+
+#### `POST /resources/carts/{cart_id}/items/`
+
+| Param Type | Param Name | Association |
+| --- | --- | --- |
+| URL Path | `cart_id` | `Cart.id` |
+
+* The request body will be similar to `CartItem` without `cart_id`.
+* The endpoint will also respond **HTTP 404** if the corresponding cart is not found.
+
+> May need a new field to store SHA1 hash on the item data for quick check.
+
+#### `DELETE /resources/carts/{cart_id}/items/{item_id}`
+
+| Param Type | Param Name | Association |
+| --- | --- | --- |
+| URL Path | `cart_id` | `Cart.id` |
+| URL Path | `item_id` | `CartItem.id` |
+
+* The endpoint will also respond **HTTP 404** if the corresponding cart is not found.
+
+### Other APIs
+
+#### `POST /resources/carts/{cart_id}/items/query`
+
+Save the cart items in batch by query.
+
+| Param Type | Param Name | Association |
+| --- | --- | --- |
+| URL Path | `cart_id` | `Cart.id` |
+
+TL;DR:
+* Respond **HTTP 202** on successfully accepting a request.
+* Use `POST /resources/cart-item-batch-jobs/{job_id}` to check the async job status.
+
+The response body for HTTP 202 will be `CartItemBatchJob`.
+
+##### Technical Considerations
+
+As this operation may require more than 30 seconds to run, the batch operation is divided into three major parts: processing requests, job updates, and job reports.
```

Comment:
Processing requests can be broken into multiple parts: handling client request, creating cart item batches, writing batches to dynamo.  Basically λ 1 to 3 in the diagram.